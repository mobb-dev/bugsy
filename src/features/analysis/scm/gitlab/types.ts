import { z } from 'zod'

export const GitlabAuthResultZ = z.object({
  access_token: z.string(),
  token_type: z.string(),
  refresh_token: z.string(),
})

export enum GitlabTokenRequestTypeEnum {
  CODE = 'code',
  REFRESH_TOKEN = 'refresh_token',
}
