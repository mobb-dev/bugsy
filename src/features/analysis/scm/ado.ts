import querystring from 'node:querystring'

import * as api from 'azure-devops-node-api'
import { z } from 'zod'

import {
  InvalidAccessTokenError,
  InvalidRepoUrlError,
  InvalidUrlPatternError,
  ReferenceType,
  RefNotFoundError,
  ScmRepoInfo,
} from './scm'
import { parseScmURL } from './urlParser'

function removeTrailingSlash(str: string) {
  return str.trim().replace(/\/+$/, '')
}

async function _getOrgsForOauthToken({ oauthToken }: { oauthToken: string }) {
  const profileZ = z.object({
    displayName: z.string(),
    publicAlias: z.string().min(1),
    emailAddress: z.string(),
    coreRevision: z.number(),
    timeStamp: z.string(),
    id: z.string(),
    revision: z.number(),
  })
  const accountsZ = z.object({
    count: z.number(),
    value: z.array(
      z.object({
        accountId: z.string(),
        accountUri: z.string(),
        accountName: z.string(),
      })
    ),
  })

  const profileRes = await fetch(
    'https://app.vssps.visualstudio.com/_apis/profile/profiles/me?api-version=6.0',
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${oauthToken}`,
      },
    }
  )
  const profileJson = await profileRes.json()
  const profile = profileZ.parse(profileJson)
  const accountsRes = await fetch(
    `https://app.vssps.visualstudio.com/_apis/accounts?memberId=${profile.publicAlias}&api-version=6.0`,
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${oauthToken}`,
      },
    }
  )
  const accountsJson = await accountsRes.json()
  const accounts = accountsZ.parse(accountsJson)
  const orgs = accounts.value
    .map((account) => account.accountName)
    .filter((value, index, array) => array.indexOf(value) === index)
  return orgs
}

function _getPublicAdoClient({ orgName }: { orgName: string }) {
  const orgUrl = `https://dev.azure.com/${orgName}`
  const authHandler = api.getPersonalAccessTokenHandler('')
  authHandler.canHandleAuthentication = () => false
  authHandler.prepareRequest = (_options) => {
    return
  }
  const connection = new api.WebApi(orgUrl, authHandler)
  return connection
}

export enum AdoTokenTypeEnum {
  PAT = 'PAT',
  OAUTH = 'OAUTH',
}

export function getAdoTokenType(token: string) {
  if (token.includes('.')) {
    return AdoTokenTypeEnum.OAUTH
  }
  return AdoTokenTypeEnum.PAT
}

export async function getAdoApiClient({
  accessToken,
  tokenOrg,
  orgName,
}: {
  accessToken: string | undefined
  tokenOrg: string | undefined
  orgName: string
}) {
  if (!accessToken || (tokenOrg && tokenOrg !== orgName)) {
    return _getPublicAdoClient({ orgName })
  }
  const orgUrl = `https://dev.azure.com/${orgName}`
  if (getAdoTokenType(accessToken) === AdoTokenTypeEnum.OAUTH) {
    const connection = new api.WebApi(orgUrl, api.getBearerHandler(accessToken))
    return connection
  }
  const authHandler = api.getPersonalAccessTokenHandler(accessToken)
  const connection = new api.WebApi(orgUrl, authHandler)
  return connection
}

export async function adoValidateParams({
  url,
  accessToken,
  tokenOrg,
}: {
  url: string | undefined
  accessToken: string | undefined
  tokenOrg: string | undefined
}) {
  try {
    if (
      !url &&
      accessToken &&
      getAdoTokenType(accessToken) === AdoTokenTypeEnum.OAUTH
    ) {
      await _getOrgsForOauthToken({ oauthToken: accessToken })
      return
    }
    let org = tokenOrg
    if (url) {
      const { owner } = parseAdoOwnerAndRepo(url)
      org = owner
    }
    if (!org) {
      throw new InvalidRepoUrlError(`invalid ADO ORG ${org}`)
    }
    const api = await getAdoApiClient({
      accessToken,
      tokenOrg,
      orgName: org,
    })
    await api.connect()
  } catch (e) {
    const error = e as {
      code?: string
      status?: number
      statusCode?: number
      response?: { status?: number; statusCode?: number; code?: string }
      description?: string
    }
    const code =
      error.code ||
      error.status ||
      error.statusCode ||
      error.response?.status ||
      error.response?.statusCode ||
      error.response?.code

    const description = error.description || `${e}`
    if (
      code === 401 ||
      code === 403 ||
      description.includes('401') ||
      description.includes('403')
    ) {
      throw new InvalidAccessTokenError(`invalid ADO access token`)
    }
    if (
      code === 404 ||
      description.includes('404') ||
      description.includes('Not Found')
    ) {
      throw new InvalidRepoUrlError(`invalid ADO repo URL ${url}`)
    }
    throw e
  }
}

export async function getAdoIsUserCollaborator({
  accessToken,
  tokenOrg,
  repoUrl,
}: {
  accessToken: string
  tokenOrg: string | undefined
  repoUrl: string
}) {
  try {
    const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
    const api = await getAdoApiClient({
      accessToken,
      tokenOrg,
      orgName: owner,
    })
    const git = await api.getGitApi()
    const branches = await git.getBranches(repo, projectName)
    if (!branches || branches.length === 0) {
      throw new InvalidRepoUrlError('no branches')
    }
    return true
  } catch (e) {
    return false
  }
}

export enum AdoPullRequestStatusEnum {
  completed = 'completed',
  active = 'active',
  all = 'all',
  abandoned = 'abandoned',
}

export const adoStatusNumberToEnumMap = {
  1: AdoPullRequestStatusEnum.active,
  2: AdoPullRequestStatusEnum.abandoned,
  3: AdoPullRequestStatusEnum.completed,
  4: AdoPullRequestStatusEnum.all,
}

export async function getAdoPullRequestStatus({
  accessToken,
  tokenOrg,
  repoUrl,
  prNumber,
}: {
  accessToken: string
  tokenOrg: string | undefined
  repoUrl: string
  prNumber: number
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  const api = await getAdoApiClient({
    accessToken,
    tokenOrg,
    orgName: owner,
  })
  const git = await api.getGitApi()
  const res = await git.getPullRequest(repo, prNumber, projectName)
  if (!res.status || res.status < 1 || res.status > 3) {
    throw new Error('bad pr status for ADO')
  }
  return adoStatusNumberToEnumMap[res.status]
}

export async function getAdoIsRemoteBranch({
  accessToken,
  tokenOrg,
  repoUrl,
  branch,
}: {
  accessToken: string
  tokenOrg: string | undefined
  repoUrl: string
  branch: string
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  const api = await getAdoApiClient({
    accessToken,
    tokenOrg,
    orgName: owner,
  })
  const git = await api.getGitApi()
  try {
    const branchStatus = await git.getBranch(repo, branch, projectName)
    if (!branchStatus || !branchStatus.commit) {
      throw new InvalidRepoUrlError('no branch status')
    }
    return branchStatus.name === branch
  } catch (e) {
    return false
  }
}

export async function getAdoRepoList({
  orgName,
  tokenOrg,
  accessToken,
}: {
  orgName: string | undefined
  tokenOrg: string | undefined
  accessToken: string
}) {
  let orgs: string[] = []
  if (getAdoTokenType(accessToken) === AdoTokenTypeEnum.OAUTH) {
    orgs = await _getOrgsForOauthToken({ oauthToken: accessToken })
  }
  if (orgs.length === 0 && !orgName) {
    throw new Error(`no orgs for ADO`)
  } else if (orgs.length === 0 && orgName) {
    orgs = [orgName as string]
  }
  const repos = (
    await Promise.allSettled(
      orgs.map(async (org) => {
        const orgApi = await getAdoApiClient({
          accessToken,
          tokenOrg,
          orgName: org,
        })
        const gitOrg = await orgApi.getGitApi()
        const orgRepos = await gitOrg.getRepositories()
        const repoInfoList = (
          await Promise.allSettled(
            orgRepos.map(async (repo) => {
              if (!repo.name || !repo.remoteUrl || !repo.defaultBranch) {
                throw new InvalidRepoUrlError('bad repo')
              }
              const branch = await gitOrg.getBranch(
                repo.name,
                repo.defaultBranch.replace(/^refs\/heads\//, ''),
                repo.project?.name
              )
              return {
                repoName: repo.name,
                repoUrl: repo.remoteUrl.replace(
                  /^[hH][tT][tT][pP][sS]:\/\/[^/]+@/,
                  'https://'
                ),
                repoOwner: org,
                repoIsPublic: repo.project?.visibility === 2, //2 is public in the ADO API
                repoLanguages: [],
                repoUpdatedAt:
                  branch.commit?.committer?.date?.toDateString() ||
                  repo.project?.lastUpdateTime?.toDateString() ||
                  new Date().toDateString(),
              }
            })
          )
        ).reduce((acc, res) => {
          if (res.status === 'fulfilled') {
            acc.push(res.value)
          }
          return acc
        }, [] as ScmRepoInfo[])
        return repoInfoList
      })
    )
  ).reduce((acc, res) => {
    if (res.status === 'fulfilled') {
      return acc.concat(res.value)
    }
    return acc
  }, [] as ScmRepoInfo[])
  return repos
}

export function getAdoDownloadUrl({
  repoUrl,
  branch,
}: {
  repoUrl: string
  branch: string
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  return `https://dev.azure.com/${owner}/${projectName}/_apis/git/repositories/${repo}/items/items?path=/&versionDescriptor[versionOptions]=0&versionDescriptor[versionType]=commit&versionDescriptor[version]=${branch}&resolveLfs=true&$format=zip&api-version=5.0&download=true`
}

export async function getAdoBranchList({
  accessToken,
  tokenOrg,
  repoUrl,
}: {
  accessToken: string
  tokenOrg: string | undefined
  repoUrl: string
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  const api = await getAdoApiClient({
    accessToken,
    tokenOrg,
    orgName: owner,
  })
  const git = await api.getGitApi()
  try {
    const res = await git.getBranches(repo, projectName)
    res.sort((a, b) => {
      if (!a.commit?.committer?.date || !b.commit?.committer?.date) {
        return 0
      }
      return (
        b.commit?.committer?.date.getTime() -
        a.commit?.committer?.date.getTime()
      )
    })
    return res.reduce((acc, branch) => {
      if (!branch.name) {
        return acc
      }
      acc.push(branch.name)
      return acc
    }, [] as string[])
  } catch (e) {
    return []
  }
}

export async function createAdoPullRequest(options: {
  accessToken: string
  tokenOrg: string | undefined
  targetBranchName: string
  sourceBranchName: string
  title: string
  body: string
  repoUrl: string
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(options.repoUrl)
  const api = await getAdoApiClient({
    accessToken: options.accessToken,
    tokenOrg: options.tokenOrg,
    orgName: owner,
  })
  const git = await api.getGitApi()
  const res = await git.createPullRequest(
    {
      sourceRefName: `refs/heads/${options.sourceBranchName}`,
      targetRefName: `refs/heads/${options.targetBranchName}`,
      title: options.title,
      description: options.body,
    },
    repo,
    projectName
  )
  return res.pullRequestId
}

export async function getAdoRepoDefaultBranch({
  repoUrl,
  tokenOrg,
  accessToken,
}: {
  repoUrl: string
  tokenOrg: string | undefined
  accessToken: string | undefined
}): Promise<string> {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  const api = await getAdoApiClient({
    accessToken,
    tokenOrg,
    orgName: owner,
  })
  const git = await api.getGitApi()
  const branches = await git.getBranches(repo, projectName)
  if (!branches || branches.length === 0) {
    throw new InvalidRepoUrlError('no branches')
  }
  const res = branches.find((branch) => branch.isBaseVersion)
  if (!res || !res.name) {
    throw new InvalidRepoUrlError('no default branch')
  }
  return res.name
}

export async function getAdoReferenceData({
  ref,
  repoUrl,
  accessToken,
  tokenOrg,
}: {
  ref: string
  repoUrl: string
  accessToken: string | undefined
  tokenOrg: string | undefined
}) {
  const { owner, repo, projectName } = parseAdoOwnerAndRepo(repoUrl)
  const api = await getAdoApiClient({
    accessToken,
    tokenOrg,
    orgName: owner,
  })
  if (!projectName) {
    throw new InvalidUrlPatternError('no project name')
  }
  const git = await api.getGitApi()
  const results = await Promise.allSettled([
    (async () => {
      const res = await git.getBranch(repo, ref, projectName)
      if (!res.commit || !res.commit.commitId) {
        throw new InvalidRepoUrlError('no commit on branch')
      }
      return {
        sha: res.commit.commitId,
        type: ReferenceType.BRANCH,
        date: res.commit.committer?.date || new Date(),
      }
    })(),
    (async () => {
      const res = await git.getCommits(
        repo,
        {
          fromCommitId: ref,
          toCommitId: ref,
          $top: 1,
        },
        projectName
      )
      const commit = res[0]
      if (!commit || !commit.commitId) {
        throw new Error('no commit')
      }
      return {
        sha: commit.commitId,
        type: ReferenceType.COMMIT,
        date: commit.committer?.date || new Date(),
      }
    })(),
    (async () => {
      const res = await git.getRefs(repo, projectName, `tags/${ref}`)
      if (!res[0] || !res[0].objectId) {
        throw new Error('no tag ref')
      }
      let objectId = res[0].objectId
      try {
        //in some cases the call to git.getRefs() returns the sha of the commit in the objectId and in some cases
        //it returns the tag object ID which we then need to call git.getAnnotatedTag() on it
        const tag = await git.getAnnotatedTag(projectName, repo, objectId)
        if (tag.taggedObject?.objectId) {
          objectId = tag.taggedObject.objectId
        }
      } catch (e) {
        /* empty */
      }
      const commitRes = await git.getCommits(
        repo,
        {
          fromCommitId: objectId,
          toCommitId: objectId,
          $top: 1,
        },
        projectName
      )
      const commit = commitRes[0]
      if (!commit) {
        throw new Error('no commit')
      }
      return {
        sha: objectId,
        type: ReferenceType.TAG,
        date: commit.committer?.date || new Date(),
      }
    })(),
  ])
  const [branchRes, commitRes, tagRes] = results
  if (tagRes.status === 'fulfilled') {
    return tagRes.value
  }
  if (branchRes.status === 'fulfilled') {
    return branchRes.value
  }
  if (commitRes.status === 'fulfilled') {
    return commitRes.value
  }
  throw new RefNotFoundError(`ref: ${ref} does not exist`)
}

export function parseAdoOwnerAndRepo(adoUrl: string) {
  adoUrl = removeTrailingSlash(adoUrl)
  const parsingResult = parseScmURL(adoUrl)

  if (!parsingResult || parsingResult.hostname !== 'dev.azure.com') {
    throw new InvalidUrlPatternError(`invalid ADO repo URL: ${adoUrl}`)
  }

  const { organization, repoName, projectName, projectPath, pathElements } =
    parsingResult
  return {
    owner: organization,
    repo: repoName,
    projectName,
    projectPath,
    pathElements,
  }
}

export async function getAdoBlameRanges() {
  return []
}

const ADO_ACCESS_TOKEN_URL = 'https://app.vssps.visualstudio.com/oauth2/token'

export enum AdoTokenRequestTypeEnum {
  CODE = 'code',
  REFRESH_TOKEN = 'refresh_token',
}

const AdoAuthResultZ = z.object({
  access_token: z.string().min(1),
  token_type: z.string().min(1),
  refresh_token: z.string().min(1),
})

export async function getAdoToken({
  token,
  adoClientSecret,
  tokenType,
  redirectUri,
}: {
  token: string
  adoClientSecret: string
  tokenType: AdoTokenRequestTypeEnum
  redirectUri: string
}) {
  const res = await fetch(ADO_ACCESS_TOKEN_URL, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: querystring.stringify({
      client_assertion_type:
        'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
      client_assertion: adoClientSecret,
      redirect_uri: redirectUri,
      assertion: token,
      grant_type:
        tokenType === AdoTokenRequestTypeEnum.CODE
          ? 'urn:ietf:params:oauth:grant-type:jwt-bearer'
          : 'refresh_token',
    }),
  })
  const authResult = await res.json()
  return AdoAuthResultZ.parse(authResult)
}
