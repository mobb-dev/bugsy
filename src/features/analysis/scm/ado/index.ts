export * from './ado'
export * from './AdoSCMLib'
export * from './constants'
export * from './types'
export * from './utils'
