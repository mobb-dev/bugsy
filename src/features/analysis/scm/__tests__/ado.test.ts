import path from 'node:path'

import chalk from 'chalk'
import * as dotenv from 'dotenv'
import { beforeAll, describe, expect, it } from 'vitest'
import { z } from 'zod'

import { getAdoClientParams, getAdoSdk, parseAdoOwnerAndRepo } from '../ado'
import { SCMLib } from '../scm'
import { ReferenceType, ScmLibScmType } from '../types'
import { env } from './env'

const TEST_ADO_REPO = 'https://dev.azure.com/mobbtest/test/_git/repo1'
dotenv.config({ path: path.join(__dirname, '../../../../../.env') })

const envVariables = z
  .object({ ADO_TEST_ACCESS_TOKEN: z.string().min(1) })
  .required()
  .parse(process.env)

type TestInput = {
  ADO_PAT: string | undefined
  PAT_ORG: string | undefined
  ADO_URL: string
  NON_EXISTING_ADO_URL: string
  EXISTING_COMMIT: string
  EXISTING_BRANCH: string
  EXISTING_BRANCH_SHA: string
  NON_EXISTING_BRANCH: string
  EXISTING_TAG: string | undefined
  EXISTING_TAG_SHA: string
}
const testNames = {
  accessTokenTest: 'accessTokenTest',
  publicRepoTest: 'publicRepoTest',
  repoWithSpacesTest: 'repoWithSpacesTest',
  publicRepoWithPat: 'publicRepoWithPat',
} as const
type TestNames = (typeof testNames)[keyof typeof testNames]

const testInputs: Record<TestNames, TestInput> = {
  accessTokenTest: {
    ADO_PAT: envVariables.ADO_TEST_ACCESS_TOKEN,
    PAT_ORG: 'mobbtest',
    ADO_URL: 'https://dev.azure.com/mobbtest/test/_git/repo1',
    NON_EXISTING_ADO_URL: 'https://dev.azure.com/mobbtest/test/_git/repo2',
    EXISTING_COMMIT: 'e619caab398d5b75621dc8aab9e480eaee5ce42d',
    NON_EXISTING_BRANCH: 'non-existing-branch',
    EXISTING_TAG: 'test-tag',
    EXISTING_BRANCH: 'main',
    EXISTING_BRANCH_SHA: 'b67eb441420675e0f107e2c1a3ba04900fc110fb',
    EXISTING_TAG_SHA: '9193050cf8c314dd52638d4bc2720b3ab48101a9',
  },
  publicRepoTest: {
    ADO_PAT: undefined,
    PAT_ORG: undefined,
    ADO_URL: 'https://dev.azure.com/mobbtest/test-public/_git/repo-public',
    NON_EXISTING_ADO_URL: 'https://dev.azure.com/mobbtest/test/_git/repo2',
    EXISTING_COMMIT: 'b67eb441420675e0f107e2c1a3ba04900fc110fb',
    EXISTING_BRANCH: 'pt',
    NON_EXISTING_BRANCH: 'non-existing-branch',
    EXISTING_TAG: 'v2023.8',
    EXISTING_BRANCH_SHA: 'b67eb441420675e0f107e2c1a3ba04900fc110fb',
    EXISTING_TAG_SHA: '5357a65e054976cd7d79b81ef3906ded050ed921',
  },
  publicRepoWithPat: {
    ADO_PAT: envVariables.ADO_TEST_ACCESS_TOKEN,
    PAT_ORG: 'mobbtest',
    ADO_URL: 'https://dev.azure.com/yhaggai/_git/hello_public',
    NON_EXISTING_ADO_URL:
      'https://dev.azure.com/mobbtest/test-public/_git/repo%20with%20spaces1',
    EXISTING_COMMIT: '51b861ec104fc5f7cf4bba564d7529d011fa5e7b',
    EXISTING_BRANCH: 'main',
    NON_EXISTING_BRANCH: 'non-existing-branch',
    EXISTING_TAG: undefined,
    EXISTING_BRANCH_SHA: '208ca7e6189b8c90bf4a9d3179f27f19a6c98940',
    EXISTING_TAG_SHA: '208ca7e6189b8c90bf4a9d3179f27f19a6c98940',
  },
  repoWithSpacesTest: {
    ADO_PAT: envVariables.ADO_TEST_ACCESS_TOKEN,
    PAT_ORG: 'mobbtest',
    ADO_URL:
      'https://dev.azure.com/mobbtest/test-public/_git/repo%20with%20spaces',
    NON_EXISTING_ADO_URL:
      'https://dev.azure.com/mobbtest/test-public/_git/repo%20with%20spaces1',
    EXISTING_COMMIT: 'd14918a74b1dd2c26726f71cb85059e63e033988',
    EXISTING_BRANCH: 'main',
    NON_EXISTING_BRANCH: 'non-existing-branch',
    EXISTING_TAG: 'test-tag',
    EXISTING_BRANCH_SHA: 'd14918a74b1dd2c26726f71cb85059e63e033988',
    EXISTING_TAG_SHA: 'd14918a74b1dd2c26726f71cb85059e63e033988',
  },
  // todo: add when we set an environment for on-prem add this test params back
  // onPremTestParams: {
  //   ADO_PAT: '*****',
  //   PAT_ORG: 'DefaultCollection',
  //   ADO_URL: 'https://test1/DefaultCollection/_git/Antony%20Test%20Project',
  //   NON_EXISTING_ADO_URL:
  //     'https://test1/DefaultCollection/_git/Antony%20Test%20Project1',
  //   EXISTING_COMMIT: 'ccf9c908745139a819c53dc296f593ab2f987728',
  //   EXISTING_BRANCH: 'main',
  //   NON_EXISTING_BRANCH: 'non-existing-branch',
  //   EXISTING_TAG: 'test-tag',
  //   EXISTING_BRANCH_SHA: 'de2e157e4d3aa4467240436fe8be7535b2bbc8c7',
  //   EXISTING_TAG_SHA: 'ccf9c908745139a819c53dc296f593ab2f987728',
  // },
} as const

let adoSdkPromise: ReturnType<typeof getAdoSdk> | undefined
function getAdoSk(adoSdkPromise: ReturnType<typeof getAdoSdk> | undefined) {
  if (!adoSdkPromise) {
    throw new Error('adoSdk not initialized')
  }
  return adoSdkPromise
}

describe.each(Object.entries(testInputs))(
  'ado reference',
  (
    testName,
    {
      ADO_PAT,
      PAT_ORG: ADO_MOBB_ORG,
      ADO_URL,
      NON_EXISTING_ADO_URL,
      EXISTING_COMMIT,
      EXISTING_BRANCH,
      NON_EXISTING_BRANCH,
      EXISTING_TAG,
      EXISTING_BRANCH_SHA,
      EXISTING_TAG_SHA,
    }
  ) => {
    beforeAll(async () => {
      adoSdkPromise = getAdoSdk(
        await getAdoClientParams({
          tokenOrg: ADO_MOBB_ORG,
          accessToken: ADO_PAT,
          url: ADO_URL,
        })
      )
    })
    it(`${chalk.green.underline.bold(testName)}: test existing repo`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      expect(
        await adoSdk.getAdoRepoDefaultBranch({
          repoUrl: ADO_URL,
        })
      ).toEqual('main')
    })
    it(`${chalk.green.underline.bold(testName)}: test non existing repo`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      await expect(() =>
        adoSdk.getAdoRepoDefaultBranch({
          repoUrl: NON_EXISTING_ADO_URL,
        })
      ).rejects.toThrow()
    })
    it(`${chalk.green.underline.bold(testName)}: test if ref is correct for commit`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      const ref = await adoSdk.getAdoReferenceData({
        repoUrl: ADO_URL,
        ref: EXISTING_COMMIT,
      })
      //date returns the current date for ADO for now as the ADO API doesn't return anything
      //so we can't test for it in the snapshot as it changes for each run
      ref.date = new Date(0)
      expect(ref.sha).toBe(EXISTING_COMMIT)
      expect(ref.type).toBe(ReferenceType.COMMIT)
    })
    it(`${chalk.green.underline.bold(testName)}: test if ref is correct for branch`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      const ref = await adoSdk.getAdoReferenceData({
        repoUrl: ADO_URL,
        ref: EXISTING_BRANCH,
      })
      expect(ref.sha).toBe(EXISTING_BRANCH_SHA)
      expect(ref.type).toBe(ReferenceType.BRANCH)
    })
    it(`${chalk.green.underline.bold(testName)}: test if ref is correct for tag`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      // todo: for some reason the tag is not found when using PAT with public repo
      if (!EXISTING_TAG) {
        return
      }
      const ref = await adoSdk.getAdoReferenceData({
        repoUrl: ADO_URL,
        ref: EXISTING_TAG,
      })
      expect(ref.sha).toBe(EXISTING_TAG_SHA)
      expect(ref.type).toBe(ReferenceType.TAG)
    })
    it(`${chalk.green.underline.bold(testName)}:test we get an error for incorrect tag`, async () => {
      const adoSdk = await getAdoSk(adoSdkPromise)
      await expect(
        adoSdk.getAdoReferenceData({
          repoUrl: ADO_URL,
          ref: NON_EXISTING_BRANCH,
        })
      ).rejects.toThrow()
    })
  }
)

const scmTestParams = {
  accessTokenConfig: {
    url: TEST_ADO_REPO,
    accessToken: env.TEST_MINIMAL_WEBGOAT_ADO_TOKEN,
    scmOrg: 'mobbtest',
  },
  // todo: add when we set an environment for on-prem add this test params back
  // onPremTestParams: {
  //   url: testInputs.onPremTestParams.ADO_URL,
  //   accessToken: testInputs.onPremTestParams.ADO_ACCESS_TOKEN,
  //   scmOrg: 'DefaultCollection',
  // },
}

describe.each(Object.entries(scmTestParams))(
  'Ado scm instance',
  (testName, { url, accessToken, scmOrg }) => {
    it(`${chalk.green.bold(testName)} should return the correct headers for basic auth type `, async () => {
      const scmLib = await SCMLib.init({
        url,
        scmType: ScmLibScmType.ADO,
        accessToken,
        scmOrg,
      })
      const authHeaders = scmLib.getAuthHeaders()
      const encodedAccessToken = Buffer.from(':' + accessToken).toString(
        'base64'
      )
      expect(authHeaders).toStrictEqual({
        authorization: `Basic ${encodedAccessToken}`,
      })
      const signedRepoUrl = await scmLib.getUrlWithCredentials()
      const { protocol, pathname, hostname } = new URL(signedRepoUrl)
      expect(signedRepoUrl).toBe(
        `${protocol}//${accessToken}@${hostname}${pathname}`
      )
      const prUrl = await scmLib.getPrUrl(1)
      expect(prUrl).toMatch(`${TEST_ADO_REPO}/pullrequest/1`)
    })
  }
)

describe('Ado check all assisoated repos', () => {
  it('should return the correct repo list', async () => {
    const scmLib = await SCMLib.init({
      url: undefined,
      scmType: ScmLibScmType.ADO,
      accessToken: env.TEST_MINIMAL_WEBGOAT_ADO_TOKEN,
      scmOrg: testInputs.accessTokenTest.PAT_ORG,
    })
    const repoList = await scmLib.getRepoList(
      testInputs.accessTokenTest.PAT_ORG
    )
    expect(repoList).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          repoIsPublic: true,
          repoLanguages: [],
          repoName: 'repo with spaces',
          repoOwner: 'mobbtest',
          repoUrl:
            'https://dev.azure.com/mobbtest/test-public/_git/repo%20with%20spaces',
        }),
        expect.objectContaining({
          repoIsPublic: false,
          repoLanguages: [],
          repoName: 'repo1',
          repoOwner: 'mobbtest',
          repoUrl: 'https://dev.azure.com/mobbtest/test/_git/repo1',
        }),
        expect.objectContaining({
          repoIsPublic: true,
          repoLanguages: [],
          repoName: 'repo-public',
          repoOwner: 'mobbtest',
          repoUrl:
            'https://dev.azure.com/mobbtest/test-public/_git/repo-public',
        }),
      ])
    )
  })
})

describe.each([
  {
    ADO_URL: 'https://dev.azure.com/mobbtest/test/_git/repo1',
    INVALID_URL: 'https://invalid.com/zj-gitlab',
    EXPECTED: {
      owner: 'mobbtest',
      repo: 'repo1',
      projectName: 'test',
      projectPath: 'mobbtest/test/_git/repo1',
      pathElements: ['mobbtest', 'test', '_git', 'repo1'],
      origin: 'https://dev.azure.com',
    },
  },
  {
    ADO_URL: 'https://dev.azure.com/mobbtest/test-public/_git/repo-public',
    INVALID_URL: 'https://invalid.com/zj-gitlab',
    EXPECTED: {
      owner: 'mobbtest',
      repo: 'repo-public',
      projectName: 'test-public',
      projectPath: 'mobbtest/test-public/_git/repo-public',
      pathElements: ['mobbtest', 'test-public', '_git', 'repo-public'],
      origin: 'https://dev.azure.com',
    },
  },
])('parsing ado url', ({ ADO_URL, INVALID_URL, EXPECTED }) => {
  it('should parse the url', () => {
    expect(parseAdoOwnerAndRepo(ADO_URL)).toStrictEqual({
      owner: EXPECTED.owner,
      repo: EXPECTED.repo,
      projectName: EXPECTED.projectName,
      projectPath: EXPECTED.projectPath,
      pathElements: EXPECTED.pathElements,
      origin: EXPECTED.origin,
    })
  })
  it('should work with trailing slash', () => {
    expect(parseAdoOwnerAndRepo(`${ADO_URL} / `)).toStrictEqual({
      owner: EXPECTED.owner,
      repo: EXPECTED.repo,
      projectName: EXPECTED.projectName,
      projectPath: EXPECTED.projectPath,
      pathElements: EXPECTED.pathElements,
      origin: EXPECTED.origin,
    })
  })
  it('fail if the url is invalid', () => {
    expect(() => parseAdoOwnerAndRepo(INVALID_URL)).toThrow()
  })
})
