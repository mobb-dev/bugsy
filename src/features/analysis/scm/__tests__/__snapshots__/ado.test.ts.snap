// Vitest Snapshot v1

exports[`ado reference > test if ref is correct for branch 1`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "b67eb441420675e0f107e2c1a3ba04900fc110fb",
  "type": "BRANCH",
}
`;

exports[`ado reference > test if ref is correct for branch 2`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "b67eb441420675e0f107e2c1a3ba04900fc110fb",
  "type": "BRANCH",
}
`;

exports[`ado reference > test if ref is correct for commit 1`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "e619caab398d5b75621dc8aab9e480eaee5ce42d",
  "type": "COMMIT",
}
`;

exports[`ado reference > test if ref is correct for commit 2`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "b67eb441420675e0f107e2c1a3ba04900fc110fb",
  "type": "COMMIT",
}
`;

exports[`ado reference > test if ref is correct for tag 1`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "9193050cf8c314dd52638d4bc2720b3ab48101a9",
  "type": "TAG",
}
`;

exports[`ado reference > test if ref is correct for tag 2`] = `
{
  "date": 1970-01-01T00:00:00.000Z,
  "sha": "5357a65e054976cd7d79b81ef3906ded050ed921",
  "type": "TAG",
}
`;

exports[`parsing ado url > should parse the url 1`] = `
{
  "owner": "mobbtest",
  "pathElements": [
    "mobbtest",
    "test",
    "_git",
    "repo1",
  ],
  "projectName": "test",
  "projectPath": "mobbtest/test/_git/repo1",
  "repo": "repo1",
}
`;

exports[`parsing ado url > should parse the url 2`] = `
{
  "owner": "mobbtest",
  "pathElements": [
    "mobbtest",
    "test-public",
    "_git",
    "repo-public",
  ],
  "projectName": "test-public",
  "projectPath": "mobbtest/test-public/_git/repo-public",
  "repo": "repo-public",
}
`;

exports[`parsing ado url > should work with trailing slash 1`] = `
{
  "owner": "mobbtest",
  "pathElements": [
    "mobbtest",
    "test",
    "_git",
    "repo1",
  ],
  "projectName": "test",
  "projectPath": "mobbtest/test/_git/repo1",
  "repo": "repo1",
}
`;

exports[`parsing ado url > should work with trailing slash 2`] = `
{
  "owner": "mobbtest",
  "pathElements": [
    "mobbtest",
    "test-public",
    "_git",
    "repo-public",
  ],
  "projectName": "test-public",
  "projectPath": "mobbtest/test-public/_git/repo-public",
  "repo": "repo-public",
}
`;
